src/rust/ensogl/lib/select/src/entry.rs

```diff
@@ -0,0 +1,310 @@
+// =================
+
+/// Padding inside entry in pixels.
+pub const PADDING:f32 = 2.0;
+/// The overall entry's height (including padding).
+pub const HEIGHT:f32 = 20.0;
+/// The text size of entry's labe.
+pub const LABEL_SIZE:f32 = 12.0;
+/// The size in pixels of icons inside entries.
+pub const ICON_SIZE:f32 = 16.0;
+/// The gap between icon and label.
+pub const ICON_LABEL_GAP:f32 = 2.0;
+
+
+
+// ===================
+// === Entry Model ===
+// ===================
+
+/// Entry id. 0 is the first entry in component.
+pub type Id = usize;
+
+/// A model on which the view bases.
+#[allow(missing_docs)]
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub struct Model {
+    pub label       : String,
+    pub highlighted : Vec<TextRange<Bytes>>,
+    #[derivative(Debug="ignore")]
+    pub icon        : Option<display::object::Any>,
+}
+
+impl Model {
+    /// Create model of simple entry with given label.
+    ///
+    /// The model won't have icon nor higlighting, but those can be set using `highlight` and
+    /// `with_icon`.
+    pub fn new(label:impl Str) -> Self {
+        Self {
+            label       : label.into(),
+            highlighted : default(),
+            icon        : default(),
+        }
+    }
+
+    /// Add highlighting to the entry and return it.
+    pub fn highlight(mut self, bytes:impl IntoIterator<Item=TextRange<Bytes>>) -> Self {
+        self.highlighted.extend(bytes.into_iter());
+        self
+    }
+
+    /// Add icon to the entry and return it.
+    pub fn with_icon(mut self, icon:impl display::Object + 'static) -> Self {
+        self.icon = Some(icon.into_any());
+        self
+    }
+}
+
+
+// === Entry Model Provider ===
+
+/// The Entry Model Provider for select component.
+///
+/// The select does not display all entries at once, instead it lazily ask for models of entries
+/// when they're about to be displayed. So setting the select content is essentially providing
+/// implementor of this trait.
+pub trait ModelProvider : Debug {
+    /// Number of all entries.
+    fn entry_count(&self) -> usize;
+
+    /// Get the model of entry with given id.
+    fn get(&self, id:Id) -> Model;
+}
+
+/// A wrapper for shared instance of some ModelProvider.
+#[derive(Clone,CloneRef,Debug,Shrinkwrap)]
+pub struct AnyModelProvider(Rc<dyn ModelProvider>);
+
+impl<T:ModelProvider + 'static> From<T> for AnyModelProvider {
+    fn from(provider: T) -> Self { Self(Rc::new(provider)) }
+}
+
+impl Default for AnyModelProvider {
+    fn default() -> Self {
+        let logger = Logger::new("EmptyModelProvider");
+        EmptyProvider{logger}.into()
+    }
+}
+
+
+// === Empty Model Provider ===
+
+/// An Entry Model Provider giving no entries.
+///
+/// This is the default provider for new select components.
+#[derive(Clone,CloneRef,Debug)]
+pub struct EmptyProvider {
+    logger : Logger,
+}
+
+impl ModelProvider for EmptyProvider {
+    fn entry_count(&self) -> usize { 0 }
+    fn get(&self, id:usize) -> Model {
```

Comment:
I think this is handling errors too early. IMO this should return Option and should have maybe another function, like `get_or_report_and_ignore` which will be automatically implemented. But basically I'm trying to always expose all getters which  expose possible errors in such a way, that the user of the API can handle them.